libraries/AP_Hott_Telem/AP_Hott_Telem.cpp

```diff
@@ -0,0 +1,358 @@
+    if (airspeed && airspeed->healthy()) {
+        msg.speed = uint16_t(airspeed->get_airspeed() * 3.6 + 0.5);
+    } else {
+        msg.speed = uint16_t(ahrs.groundspeed() * 3.6 + 0.5);
+    }
+
+    send_packet((const uint8_t *)&msg, sizeof(msg));
+}
+
+void AP_Hott_Telem::GPS_to_DMS(float decimal, uint8_t &sign, uint16_t &dm, uint16_t &sec) const
+{
+    sign = decimal>=0?0:1;
+    decimal = fabsf(decimal);
+    uint8_t deg = uint16_t(decimal);
+    uint8_t min = uint16_t((decimal - deg) * 60);
+    dm = deg*100 + min;
+    sec = (decimal - (deg + min/60.0)) * 3600;
+}
+
+/*
+  send GPS packet
+ */
+void AP_Hott_Telem::send_GPS(void)
+{
+    // GPS message
+    struct PACKED {
+        uint8_t start_byte = 0x7c;    //#01 constant value 0x7c
+        uint8_t gps_sensor_id = 0x8a; //#02 constant value 0x8a
+        uint8_t warning_beeps;
+        uint8_t sensor_id = 0xA0;     //#04 constant (?) value 0xa0
+        uint16_t alarm; //#05
+        uint8_t flight_direction;     //#07 flight direction in 2 degreees/step (1 = 2degrees);
+        uint16_t gps_speed_kmh;       //#08 km/h
+        uint8_t pos_NS;               //#10 north = 0, south = 1
+        uint16_t pos_NS_dm;           //#11 degree minutes
+        uint16_t pos_NS_sec;          //#13 position seconds
+        uint8_t pos_EW;               //#15 east = 0, west = 1
+        uint16_t pos_EW_dm;           //#16 degree minutes
+        uint16_t pos_EW_sec;          //#18 position seconds
+        uint16_t home_distance;       //#20 meters
+        uint16_t altitude;            //#22 meters. Value of 500 = 0m
+        uint16_t climbrate;           //#24 m/s 0.01m/s resolution. Value of 30000 = 0.00 m/s
+        uint8_t climbrate3s;          //#26 climbrate in m/3s resolution, value of 120 = 0 m/3s
+        uint8_t gps_satelites;        //#27 sat count
+        uint8_t gps_fix_char;         //#28 GPS fix character. display, 'D' = DGPS, '2' = 2D, '3' = 3D, '-' = no fix
+        uint8_t home_direction;       //#29 direction from starting point to Model position (2 degree steps)
+        int16_t vel_north;            //#30 velocity north mm/s
+        uint8_t speed_acc;            //#32 speed accuracy cm/s
+        uint8_t gps_time_h;           //#33 UTC time hours
+        uint8_t gps_time_m;           //#34 UTC time minutes
+        uint8_t gps_time_s;           //#35 UTC time seconds
+        uint8_t gps_time_hs;          //#36 UTC time 0.01s units
+        int16_t vel_east;             //#37 velocity east mm/s
+        uint8_t horiz_acc;            //#39 horizontal accuracy dm
+        uint16_t msl_altitude;        //#40 mean sea level altitude
+        uint8_t gps_fix_char2;        //#42 GPS fix character2??. display, 'D' = DGPS, '2' = 2D, '3' = 3D, '-' = no fix
+        uint8_t version;       //#43  0: GPS Graupner #33600
+        uint8_t status1;
+        uint8_t status2;
+        uint8_t stop_byte = 0x7d;
+    } msg {};
+
+    AP_GPS &gps = AP::gps();
```

Comment:
You're taking information from this object after releasing the semaphore.

Reply:
when putting a uint32_t or less it is atomic anyway